### PR TITLE
dolt diff --summary

### LINKED ISSUE #77: dolt diff --summary
Some diffs are too big for human inspection. I want a dolt diff --summary that tells me the number of:

Rows Unmodified
Rows Added
Rows Deleted
Rows Modified
Cells Modified

for each table.

### PR DESCRIPTION
**Pull request by tbantle22**
Example output using [Liquidata/tatoeba-sentence-translations](https://www.dolthub.com/repositories/Liquidata/tatoeba-sentence-translations):
```
$ dolt diff --summary rnfm50gmumlettuebt2latmer617ni3t

diff --dolt a/sentences b/sentences
--- a/sentences @ gd1v6fsc04k5676c105d046m04hla3ia
+++ b/sentences @ 2ttci8id13mijhv8u94qlioqegh7lgpo
7,800,102 Rows Unmodified (99.99%)
15,030 Rows Added (0.19%)
108 Rows Deleted (0.00%)
960 Rows Modified (0.01%)
1,888 Cells Modified (0.00%)
(7,801,170 Entries vs 7,816,092 Entries)

diff --dolt a/translations b/translations
--- a/translations @ p2355o6clst8ssvr9jha2bfgqbrstkmm
+++ b/translations @ 62ri8lmohbhs1mc01m9o4rbvj6rbl8ee
5,856,845 Rows Unmodified (90.91%)
468,173 Rows Added (7.27%)
578,242 Rows Deleted (8.98%)
7,626 Rows Modified (0.12%)
7,626 Cells Modified (0.06%)
(6,442,713 Entries vs 6,332,494 Entries)
```
Fixes #77

### CODE REVIEW

File: go/libraries/doltcore/diff/diff_summary.go [hunk 1]
```diff
@@ -0,0 +1,338 @@
+// Copyright 2019 Liquidata, Inc.
```
Comment (tbantle22) [inline go/libraries/doltcore/diff/diff_summary.go]: I copied a lot of this from noms - should I keep this file or make these changes [here](https://github.com/liquidata-inc/dolt/blob/master/go/store/diff/summary.go)?

File: go/libraries/doltcore/diff/diff_summary.go [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+			singular = "Field"
+			plural = "Fields"
+		case types.MapKind:
+			singular = "Entry"
+			plural = "Entries"
+		default:
+			singular = "Value"
+			plural = "Values"
+		}
+	}
```
Comment (tbantle22) [inline go/libraries/doltcore/diff/diff_summary.go]: Is this part necessary if `Summary` is only being used for table diffs and the `rowData` getting passed in is a `Map`? It's copied from noms

File: go/libraries/doltcore/diff/diff_summary.go (hunk 1 above)
Comment (bheni) [reply go/libraries/doltcore/diff/diff_summary.go]: I believe this is correct.  If it's not there is a check that happens, and jenkins will fail your build.

File: go/libraries/doltcore/diff/diff_summary.go (hunk 2 above)
Comment (bheni) [reply go/libraries/doltcore/diff/diff_summary.go]: if you change your function signature to use types.Map for v1 and v2 then you ensure the calling code only passes maps.  Then you can kill a lot of code such as diffSummaryList, diffSummarySet, etc.

File: go/libraries/doltcore/diff/diff_summary.go [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+
+	humanize "github.com/dustin/go-humanize"
+
+	"github.com/liquidata-inc/dolt/go/cmd/dolt/cli"
+	"github.com/liquidata-inc/dolt/go/cmd/dolt/errhand"
+	"github.com/liquidata-inc/dolt/go/store/atomicerr"
+	"github.com/liquidata-inc/dolt/go/store/datas"
+	"github.com/liquidata-inc/dolt/go/store/diff"
+	"github.com/liquidata-inc/dolt/go/store/types"
+	"github.com/liquidata-inc/dolt/go/store/util/status"
```
[the remark this answers is not in the document]
Comment (tbantle22) [reply go/libraries/doltcore/diff/diff_summary.go]: Moved `Summary` and `reportChanges` functions to `cmd` [aa96427](https://github.com/liquidata-inc/dolt/pull/105/files#diff-8afdef1a3c9e0c5f667365e20b8a046fR575-R642)

File: go/libraries/doltcore/diff/diff_summary.go [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+	}
+
+	var rp atomic.Value
+	var verr errhand.VerboseError
+	ae := atomicerr.New()
+	ch := make(chan diffSummaryProgress)
+	go func() {
+		defer close(ch)
+		defer func() {
+			if r := recover(); r != nil {
```
[the remark this answers is not in the document]
Comment (tbantle22) [reply go/libraries/doltcore/diff/diff_summary.go]: Took out the recover calls in summary functions in [cmd ](https://github.com/liquidata-inc/dolt/pull/105/files#diff-8afdef1a3c9e0c5f667365e20b8a046fR580) and [library](https://github.com/liquidata-inc/dolt/pull/105/files#diff-f7ee563c78a9fb4951bf98eae945d4bcR53)

File: go/libraries/doltcore/diff/diff_summary.go [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+				rp.Store(r)
+			}
+		}()
+		verr = diffSummary(ctx, ae, ch, v2, v1)
```
[the remark this answers is not in the document]
Comment (tbantle22) [reply go/libraries/doltcore/diff/diff_summary.go]: Fixed and summary functions in `libraries` don't return an error anymore

File: go/libraries/doltcore/diff/diff_summary.go [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+			case types.ListKind:
+				verr = diffSummaryList(ctx, ae, ch, v1.(types.List), v2.(types.List))
+			case types.MapKind:
+				verr = diffSummaryMap(ctx, ae, ch, v1.(types.Map), v2.(types.Map))
+			case types.SetKind:
+				verr = diffSummarySet(ctx, ae, ch, v1.(types.Set), v2.(types.Set))
+			case types.StructKind:
+				verr = diffSummaryStructs(ae, ch, v1.(types.Struct), v2.(types.Struct))
+			default:
+				return errhand.BuildDError("Unrecognized type in diff function").Build()
```
[the remark this answers is not in the document]
Comment (tbantle22) [reply go/libraries/doltcore/diff/diff_summary.go]: Ended up removing switch statement altogether [aa96427](https://github.com/liquidata-inc/dolt/pull/105/files#diff-f7ee563c78a9fb4951bf98eae945d4bcR36)

File: go/libraries/doltcore/diff/diff_summary.go (hunk 2 above)
Comment (tbantle22) [reply go/libraries/doltcore/diff/diff_summary.go]: Fixed and removed all those functions/code having to do with types

File: go/libraries/doltcore/diff/diff_summary.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+type diffFunc func(changeChan chan<- types.ValueChanged, stopChan <-chan struct{})
+
+type DiffSummaryProgress struct {
+	Adds, Removes, Changes, CellChanges, NewSize, OldSize uint64
+}
+
+// Summary reports a summary of diff changes between two values
+func Summary(ctx context.Context, ae *atomicerr.AtomicError, ch chan DiffSummaryProgress, v1, v2 types.Map) {
+	if !v1.Equals(v2) {
+		if diff.ShouldDescend(v1, v2) {
```
Comment (bheni) [inline go/libraries/doltcore/diff/diff_summary.go]: Can ShouldDescend ever be false for 2 maps?  If not I'd probably just remove the if statement, and put the logic of diffSummaryValueChanged in here directly.

Comment (tbantle22) [reply go/libraries/doltcore/diff/diff_summary.go]: No it shouldn't - took out the if statement and combined those functions [4da04bc](https://github.com/liquidata-inc/dolt/pull/105/files#diff-f7ee563c78a9fb4951bf98eae945d4bcR32)

**Pull request merged**